File: frontend/src/api/recipients.ts
import { api } from "./client";
import type {
  CareerRecord,
  GenerateDocumentResponse,
  MeritContent,
  PreviousAward,
  Recipient,
  RecipientDetail,
  URLExtractResponse,
} from "../types";

export const createRecipient = (caseId: string, payload: Partial<Recipient>) =>
  api.post<RecipientDetail>(`/api/award-cases/${caseId}/recipients`, payload).then(r => r.data);

export const getRecipient = (id: string) =>
  api.get<RecipientDetail>(`/api/recipients/${id}`).then(r => r.data);

export const updateRecipient = (id: string, payload: Partial<Recipient>) =>
  api.patch<RecipientDetail>(`/api/recipients/${id}`, payload).then(r => r.data);

export const deleteRecipient = (id: string) =>
  api.delete(`/api/recipients/${id}`).then(r => r.data);

export const upsertMerit = (id: string, payload: Partial<MeritContent>) =>
  api.put<MeritContent>(`/api/recipients/${id}/merit-content`, payload).then(r => r.data);

export const generateMerit = (id: string, payload: { keywords: string[]; activity_summary?: string }) =>
  api.post<MeritContent>(`/api/recipients/${id}/generate-merit`, payload).then(r => r.data);

export const generatePdf = (id: string) =>
  api.post<GenerateDocumentResponse>(`/api/recipients/${id}/generate-pdf`).then(r => r.data);

export const addCareer = (id: string, payload: Partial<CareerRecord>) =>
  api.post<CareerRecord>(`/api/recipients/${id}/career-records`, payload).then(r => r.data);

export const deleteCareer = (recordId: string) =>
  api.delete(`/api/career-records/${recordId}`).then(r => r.data);

export const addPreviousAward = (id: string, payload: Partial<PreviousAward>) =>
  api.post<PreviousAward>(`/api/recipients/${id}/previous-awards`, payload).then(r => r.data);

export const deletePreviousAward = (recordId: string) =>
  api.delete(`/api/previous-awards/${recordId}`).then(r => r.data);

export const extractFromUrl = (url: string) =>
  api.post<URLExtractResponse>(`/api/extract-from-url`, { url }).then(r => r.data);

File: backend/app/api/recipients.py
"""표창 대상자 API"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import HTMLResponse
from sqlalchemy.orm import Session

from .. import models, schemas
from ..database import get_db
from ..services import pdf_generator
from .deps import get_case_or_404, get_recipient_or_404

router = APIRouter(tags=["recipients"])


@router.post(
    "/api/award-cases/{case_id}/recipients",
    response_model=schemas.RecipientDetail,
)
def create_recipient(case_id: str, payload: schemas.RecipientCreate, db: Session = Depends(get_db)):
    case = get_case_or_404(db, case_id)
    data = payload.model_dump()
    if not data.get("sequence_no"):
        data["sequence_no"] = (len(case.recipients) + 1) if case.recipients else 1
    recipient = models.Recipient(award_case_id=case.id, **data)

    # 빈 merit_content 자동 생성
    recipient.merit_content = models.MeritContent()
    db.add(recipient)
    db.commit()
    db.refresh(recipient)
    return schemas.RecipientDetail.model_validate(recipient)


@router.get("/api/recipients/{recipient_id}", response_model=schemas.RecipientDetail)
def get_recipient(recipient_id: str, db: Session = Depends(get_db)):
    r = get_recipient_or_404(db, recipient_id)
    return schemas.RecipientDetail.model_validate(r)


@router.patch("/api/recipients/{recipient_id}", response_model=schemas.RecipientDetail)
def update_recipient(recipient_id: str, payload: schemas.RecipientUpdate, db: Session = Depends(get_db)):
    r = get_recipient_or_404(db, recipient_id)
    for k, v in payload.model_dump(exclude_unset=True).items():
        setattr(r, k, v)
    db.commit()
    db.refresh(r)
    return schemas.RecipientDetail.model_validate(r)


@router.delete("/api/recipients/{recipient_id}")
def delete_recipient(recipient_id: str, db: Session = Depends(get_db)):
    r = get_recipient_or_404(db, recipient_id)
    db.delete(r)
    db.commit()
    return {"ok": True}


@router.get("/api/recipients/{recipient_id}/preview", response_class=HTMLResponse)
def preview_html(recipient_id: str, db: Session = Depends(get_db)):
    """공적조서 HTML 미리보기"""
    r = get_recipient_or_404(db, recipient_id)
    if not r.award_case:
        raise HTTPException(status_code=400, detail="표창 건 정보가 없습니다")
    html = pdf_generator.render_html_for_preview(r.award_case, r)
    return HTMLResponse(content=html)
